### ISSUE TITLE
PG types without timezone being parsed into time.Times with.

### ISSUE THREAD

**Issue by e-dard** [lib/pq#329]
Hi,

I briefly discussed this with @johto in `#postgresql`.

Currently, when a time-based PG type that doesn't have timezone information is parsed into a Go type, it's parsed into a `time.Time` that does contain timezone information. Specifically, the resulting `time.Time` contains a single `zone` in the `Location`, while the zero-value for a `time.Location` would contain an empty slice of `zones`, i.e., the absence of any timezones.

Given the `PG` value `'2015-01-13'::date` it should be equivalent to the following `time.Time`:

```go
time.Time{
        sec: 63556704000, 
        nsec: 0, 
        // zero value for a *time.Location
        loc: &time.Location{
                name: "", 
                zone: []time.zone(nil), 
                tx: []time.zoneTrans(nil), 
                cacheStart: 0, 
                cacheEnd: 0, 
                cacheZone: (*time.zone)(nil),
         },
}
```

Currently, `lib/pq` would return the following `time.Time`:

```go
time.Time{
    sec: 63556704000, 
    nsec: 0, 
    loc: &time.Location{
        name: "", 
        zone: []time.zone{
            time.zone{
                name: "", 
                offset: 0, 
                isDST: false,
            },
        }, 
        tx: []time.zoneTrans{
            time.zoneTrans{
                when: -9223372036854775808, 
                index: 0x0, 
                isstd: false, 
                isutc: false},
            },
        },
        cacheStart: -9223372036854775808, 
        cacheEnd: 9223372036854775807, 
        cacheZone:(*time.zone)(0xc20801eaa0),
    },
}
```

I'm don't think this is correct because the `time.Location` produced by `lib/pq` contains zone information, even if the zone is itself the zero-value.

I would be happy to put a fix in for this, but wanted to see if there was support for that idea or not?  I've added a PR with a test case highlighting the issue.

**Comment by johto** [comment]
I think what we'd need to consider here are:

  1. How many applications would changing this realistically break
  2. What are the practical gains (meaning "outside of test cases") of changing this?  Are there, for example, any calculations which give the wrong (or at least a dubious) answer with the time.Time value returned now?

I can't claim to know the time package well enough to answer the second question without any research.

**Comment by johto** [comment]
I've given this some more thought, and if this doesn't break any calculations, I think we should fix this.  It might break someone's test cases, but I think doing the correct think going forward is more important than that.

**Comment by e-dard** [comment]
@johto sorry for the late reply to your original comment.

I agree (obviously) :smile: 

Shall I try and put together a fix for this issue and then go from there?

**Comment by johto** [comment]
I think the first thing would be to research to see if there are any cases where the user would see a difference in behavior.

**Comment by cbandy** [comment]
Maybe `time.Time` isn't the appropriate Go type for these Postgres types.

http://golang.org/pkg/time/#Time:
> A Time represents an instant in time with nanosecond precision.
> Each Time has associated with it a Location...

http://golang.org/pkg/time/#Date:
> Date panics if loc is nil.

I don't have access to check right now, but I wonder how Postgres handles comparing `date` to `timestamp` (with and without time zone) values?

**Comment by johto** [comment]
> Maybe time.Time isn't the appropriate Go type for these Postgres types.

Yeah, it's not really designed for this, but it's very very close.

> http://golang.org/pkg/time/#Time:
> 
> A Time represents an instant in time with nanosecond precision.
> Each Time has associated with it a Location...
> http://golang.org/pkg/time/#Date:
> 
> Date panics if loc is nil.

Yeah, but the idea is to use the "zero" location instead of a nil location.

> I don't have access to check right now, but I wonder how Postgres handles comparing date to timestamp (with and without time zone) values?

The date is converted to the appropriate time first.

**Comment by cbandy** [comment]
`with time zone ` is probably session timezone then?

I also forgot that users get to scan into whatever they want. If they want a `time.Time`, then I think it should either have the [backend] session timezone or the [go runtime] local timezone.

**Comment by e-dard** [comment]
>> http://golang.org/pkg/time/#Time:
>>
>> A Time represents an instant in time with nanosecond precision.
>> Each Time has associated with it a Location...
 http://golang.org/pkg/time/#Date:

>>Date panics if loc is nil.

>Yeah, but the idea is to use the "zero" location instead of a nil location.


Exactly, so in a test if you were say asserting you got a certain `DATE` type back from PG you would use the following to create the appropriate `time.Time`:

```go
expected := time.Date(2015, 2, 13, 0, 0, 0, 0, &time.Location{})
```

while currently, you have to do:

```go
time.Date(2015, 2, 13, 0, 0, 0, 0, time.FixedZone("", 0))
```

According to the time package [documentation](http://golang.org/src/time/zoneinfo.go?s=2523:2572#L79) `FixedZone`:

```go
// FixedZone returns a Location that always uses
// the given zone name and offset (seconds east of UTC).
```

I must admit, at this point in time I'm not entirely sure what the use-case for `FixedZone` actually is, but the semantics don't match up with *the absence of a time zone*, which I'm arguing `&time.Location` (and thus the zero-value for any embedded zone information, does.

**Comment by cbandy** [comment]
> I'm not entirely sure what the use-case for FixedZone actually is

http://golang.org/pkg/time/#Parse uses `FixedZone` when the input offset is not local.

**Comment by johto** [comment]
> with time zone is probably session timezone then?
> 
> I also forgot that users get to scan into whatever they want. If they want a time.Time, then I think it should either have the [backend] session timezone or the [go runtime] local timezone.

I think our behavior when scanning a `timestamp with time zone` into a time.Time is correct already.  The problem is that we claim that the time.Time when scanned from a `timestamp without time zone` has a UTC offset of 0.  But we don't know what the times represent, so it's a lie.

**Comment by johto** [comment]
Hmm..  But this doesn't look correct either:

```
    t := time.Date(2015, 2, 13, 0, 0, 0, 0, &time.Location{})
    fmt.Println(t)

% go run test.go
2015-02-13 00:00:00 +0000 UTC
```

.. UTC?  That seems to be even more wrong than the "+0000 +0000" we currently return.

**Comment by e-dard** [comment]
I think that's because of the `fmt.Stringer` [implementation](http://golang.org/src/time/zoneinfo.go#L66) for a `time.Location`.

**Comment by johto** [comment]
Yeah.  Thing is, they both behave as if they're UTC (or at least something with 0 offset from UTC).  The difference is that what we currently return doesn't claim to be UTC, and I think that's less wrong in this case.  I can't really see any other differences here *unless* I use reflection to look into the time.Time struct itself, and that's just implementation details in my mind.

I think both behaviors are wrong, but the one we currently implement is less so.

**Comment by cbandy** [comment]
On the backend, values without zone information *act like* they are in the zone of the current session:

```sql
pqgotest=> SHOW timezone;
    TimeZone     
-----------------
 America/Chicago
(1 row)

pqgotest=> SELECT a, b, a = b, a < b, a > b FROM (VALUES ('2015-02-13'::date, '2015-02-13T00:00:00+00:00'::timestamptz)) AS x(a,b);
     a      |           b            | ?column? | ?column? | ?column? 
------------+------------------------+----------+----------+----------
 2015-02-13 | 2015-02-12 18:00:00-06 | f        | f        | t
(1 row)

pqgotest=> SELECT a, b, a = b, a < b, a > b FROM (VALUES ('2015-02-13T04:05:06'::timestamp, '2015-02-13T04:05:06+00:00'::timestamptz)) AS x(a,b);
          a          |           b            | ?column? | ?column? | ?column? 
---------------------+------------------------+----------+----------+----------
 2015-02-13 04:05:06 | 2015-02-12 22:05:06-06 | f        | f        | t
(1 row)
```

However, it is possible to convert from one zone to another on the backend *and discard the zone information*:

```sql
pqgotest=> SHOW timezone;
    TimeZone     
-----------------
 America/Chicago
(1 row)

pqgotest=> SELECT a, b, a AT TIME ZONE 'America/New_York', b AT TIME ZONE 'America/New_York' FROM (VALUES ('2015-02-13T04:05:06'::timestamp, '2015-02-13T04:05:06+00:00'::timestamptz)) AS x(a,b);
          a          |           b            |        timezone        |      timezone       
---------------------+------------------------+------------------------+---------------------
 2015-02-13 04:05:06 | 2015-02-12 22:05:06-06 | 2015-02-13 03:05:06-06 | 2015-02-12 23:05:06
(1 row)
```

The short answer is, we are entirely incapable of discerning the correct offset to apply when the backend does not supply it.

>> Maybe time.Time isn't the appropriate Go type for these Postgres types.

> Yeah, it's not really designed for this, but it's very very close.

I think it is not close. Given `'2015-02-13 04:05:06'`, how should the `Time` object respond to `Unix()` or `Local()`?

The interface on `Time` is very wide and I don't think we can populate the struct correctly for every use case.

**Comment by johto** [comment]
On 2015-02-08 9:41 PM, Chris Bandy wrote:
> The short answer is, we are entirely incapable of discerning the correct offset to apply when the backend does not supply it.

There is no correct offset in that case anymore.  The timezone 
information is lost.

>>> Maybe time.Time isn't the appropriate Go type for these Postgres types.
>
>> Yeah, it's not really designed for this, but it's very very close.
>
> I think it is not close.

Yeah, sorry, you're right.  I was confused about how .In() works; it's 
actually the reverse operation we'd want in this case, i.e. "this is the 
local time in this time zone, what's the absolute time?"

> Given `'2015-02-13 04:05:06'`, how should the `Time` object respond to `Unix()` or `Local()`?

Ideally, it shouldn't.  Neither of those questions are meaningful.

> The interface on `Time` is very wide and I don't think we can populate the struct correctly for every use case.

Yeah.

**Comment by cbandy** [comment]
`time.Time` was added as a valid `driver.Value` in https://github.com/golang/go/commit/bf734d62d8210b3030757522c3e9ff581457daa4. Neither the commit, its comments nor originating issue offer any explanation for its addition. The code and comments are identical in Go 1.4.1.

There are some threads in #golang-nuts, but [the summary](https://groups.google.com/d/topic/golang-nuts/QCAao8PUnpk/discussion
) is:

> There is no such thing as "correct date/time handling" with the current package.

So, it is up to us to document how *we* have chosen to interpret/return things for [all Postgres date/time types](http://www.postgresql.org/docs/current/static/datatype-datetime.html).

---

Regarding the offset/zone/location representation of `date`, `time`, and `timestamp without time zone` types:

When returned in a `time.Time` instance, I think it is most accurate to say that **the `Location` is entirely unknown and a user should not expect nor rely on anything related to it**. Unfortunately, this includes the `fmt.Stringer` interface, `Format` layouts for time zone, `encoding` interfaces, arithmetic methods, etc. Nearly every method would be useless.

I feel we are compelled instead to assign a `Location` that puts the `time.Time` in a functional and predictable state. **What functionality do we want?**

---

1\. Presently, we return these types with a `FixedZone("", 0)`.

This deviates from behavior in the `time` package, which consistently uses [`FixedZone()` when there *is* zone information](https://github.com/golang/go/blob/go1.1/src/pkg/time/format.go#L936-L967). In fact, `FixedZone("", 0)` exactly what `Parse()` returns when there is a zero offset with no name, e.g.

```shell
TZ="America/Chicago" gore 'p time.Parse(time.RFC3339, "2001-01-01T01:01:01+00:00")'
```

This probably isn't a good sentinel value.

---

2\. The OP's expectation is that the `Location` is zero on a `time.Time` returned from these types.

The zero value is sometimes the right choice to indicate "no information." However, while the zero `Location` has no name, it is effectively [the UTC zone](https://github.com/golang/go/blob/go1.1/src/pkg/time/zoneinfo.go#L48-L54).

I'm okay with these values acting like they are in the UTC zone. In fact, when there is no zone information, [`Parse()` returns `time.UTC`](http://golang.org/pkg/time/#Parse). If we are okay with the UTC zone, should we just use the UTC location?

**Comment by jamie-stackhouse** [comment]
From my recent experience with this, any time stored in Postgres without TZ info will come back in Go as a (+|-)0000 TZ when FixedZone is used. http://golang.org/src/time/format.go#L572

Therefore, I'm with @cbandy, the only change that I feel would be valid is setting the location to UTC, as the behaviour of Go would appear to be to assume UTC when no information is available.

**Comment by pkieltyka** [comment]
@jamie-stackhouse that sounds consistent with Go's behavior and makes sense to me too

**Comment by rtyer** [comment]
Any workarounds for this?  We're using timestamp w/o timezone given everything is UTC to save on space.

**Comment by johto** [comment]
> We're using timestamp w/o timezone given everything is UTC to save on space.

Uh.  "timestamp" and "timestamp with time zone" take exactly the same amount of space on disk.  The actual time zone is not stored.

> Any workarounds for this?

Workaround for what, exactly?

**Comment by rtyer** [comment]
Yeah, disregard my previous comment entirely.  My issue was unrelated to this, in any case.  Thanks for the quick response.

**Comment by taylorchu** [comment]
I stored a time.Time in UTC, and got back a different time.Time according to reflect.DeepEqual. Hopefully this issue will be fixed soon.

**Comment by johto** [comment]
> I stored a time.Time in UTC, and got back a different time.Time according to reflect.DeepEqual.

Storing UTC times in a `timestamp without time zone` column is a bad idea.  You're just poorly reimplementing what `timestamp with time zone` already does out-of-the-box.

> Hopefully this issue will be fixed soon.

Does anyone care enough to suggest a patch?

**Comment by jbrodriguez** [comment]
> Storing UTC times in a timestamp without time zone column is a bad idea. You're just poorly reimplementing what timestamp with time zone already does out-of-the-box.

Hi johto, I assume you say that because then you wouldn't know _where_ that moment happened.

But if you have a separate timezone column, storing for instance 'America/New York', then you'd have the complete picture.

Are you saying that it would be better to just use the with time zone column and forgo a separte tz column ?

**Comment by johto** [comment]
> I assume you say that because then you wouldn't know where that moment happened.

You don't know that in either case.  `timestamp with time zone` does not store the time zone.

> But if you have a separate timezone column, storing for instance 'America/New York', then you'd have the complete picture.
> 
> Are you saying that it would be better to just use the with time zone column and forgo a separte tz column ?

That's orthogonal to what I commented on.  The original comparison was `timestamp without time zone` in UTC versus `timestamp with time zone`.  Neither know what the original time zone was.  Whether to store the original time zone or not is a completely different question.  In my experience it's usually not an interesting piece of evidence.

**Comment by jbrodriguez** [comment]
Thanks for the reply.

I'm not finding much use for a separate tz either, but it's handy if someone later asks .. hey, what's the local time for that record?, something like that

**Comment by taylorchu** [comment]
@johto 
> Storing UTC times in a timestamp without time zone column is a bad idea. You're just poorly reimplementing what timestamp with time zone already does out-of-the-box.

I am not sure the decision here was. Postgres converts any time to UTC and then stores it. Therefore it  might be better if pq returns time with location and timezone just like `time.Now().UTC()`.

**Comment by johto** [comment]
> I am not sure the decision here was. Postgres converts any time to UTC and then stores it. Therefore it might be better if pq returns time with location and timezone just like time.Now().UTC().

I'm not sure we're talking about the same thing here.  "timestamp without time zone" does no conversion all; it wouldn't know how to.

I still think that it's stupid to say the time is UTC, but then on the other hand it's stupid to say that it's +00 as well.  We should probably just be stupid in the same way Go itself is, and return a UTC time.Time instead of a +00 time.Time.

**Comment by j16sdiz** [comment]
I think we should use the session time zone.

Those who use  "timestamp without time zone" are either:

1. Don't care about the time zone; or
2. Have a separate column for location/zone information.

For the first case, local time would be most usable. 
For the second case, setup the session timezone should trivial compare to the custom zone handling they are doing.

**Comment by johto** [comment]
> I think we should use the session time zone.

No, that would be a huge backwards compatibility breakage.

Besides, I don't think that would be the correct thing to do even if we didn't have users to worry about.  There's no reason to assume that the times are in any specific time zone, or even that they're all in the same time zone.

**Comment by arekkas** [comment]
I have a similar issue, where I'm storing a `time.Time` value (created with `time.Now()`) into a pg column `timestamp with time zone NOT NULL` and `DeepEqual` fails, because the time values are different:

```
&{code 60 scope http://localhost/ state 2015-10-19 13:54:44.2847179 +0200 CEST <nil>}

&{code 60 scope http://localhost/ state 2015-10-19 11:54:44.284718 +0000 UTC <nil>}
```

Also there seems to be some rounding going on.

**Comment by cbandy** [comment]
@arekkas `time.Time` has nanosecond precision while Postgres timestamps have only microsecond precision.

**Comment by arekkas** [comment]
@cbandy thank's for the clarification :) What I overlooked is that the time is converted as it should be, so I think everything looks legit.

**Comment by taylorchu** [comment]
@johto 

I found a good example in `time.Parse` that might back the decision for having UTC as dummy default.

>  In the absence of a time zone indicator, Parse returns a time in UTC.

**Comment by icholy** [comment]
> I still think that it's stupid to say the time is UTC, but then on the other hand it's stupid to say that it's +00 as well. We should probably just be stupid in the same way Go itself is, and return a UTC time.Time instead of a +00 time.Time.

Exactly. 

My issue with this is that if you format a time with a layout that includes the timezone, you can't parse the string back using the same layout. 

``` go
var (
  t = time.Date(0, 0, 0, 0, 0, 0, 0, time.FixedZone("", 0))
  s = t.Format(time.RFC1123)
)

_, err := time.Parse(time.RFC1123, s)

fmt.Println(err)
```

http://play.golang.org/p/kIms1Bo8FC

**Comment by AlekSi** [comment]
> I still think that it's stupid to say the time is UTC, but then on the other hand it's stupid to say that it's +00 as well. We should probably just be stupid in the same way Go itself is, and return a UTC time.Time instead of a +00 time.Time.

+1. Will you accept a patch for that?

**Comment by albertsundjaja** [comment]
My app currently has problem with this, I could implement my own fix, however, has there been any official fix/recommended solution for this?

**Comment by discordianfish** [comment]
I come back to this issue again and again. Can anyone confirm that a patch to UTC would be acceptable?

**Comment by johto** [comment]
> I come back to this issue again and again. Can anyone confirm that a patch to UTC would be acceptable?

Why is this an issue?

**Comment by ainar-g** [comment]
@johto As others have said before, it makes comparing structures with `reflect.DeepEqual` impossible because of the timezone.

**Comment by thrasher-redhat** [comment]
This is still an issue whenever using and comparing dates.  

`2018-06-05 00:00:00 +0000 UTC` and `2018-06-05 00:00:00 +0000` are the same time, but aren't equal.

Currently having to .UTC() every date being read from the database that could be used for comparisons.

**Comment by tomwassing** [comment]
Any update on this issue?

**Comment by SamuelTissot** [comment]
I hope this helps
I was tired of always casting to UTC and truncating so I made a little type that does just that.

now I can use `reflect.DeepEqual` 

https://github.com/SamuelTissot/sqltime

**Comment by akshayjshah** [comment]
I pulled in the test from #329 and opened #956 to address this issue. I think the proposed fix incorporates the past few years of discussion.

**Comment by voodoo-dn** [comment]
Pulled last version v1.7.0
The same problem in my tests

```
- Birthday: (time.Time) 1980-01-21 00:00:00 +0000 UTC,
+ Birthday: (time.Time) 1980-01-21 00:00:00 +0000 +0000
```

**Comment by ivothgle** [comment]
@arp242 
Hi,

Continuing the discussion from https://github.com/lib/pq/issues/1174#issuecomment-3829113076, thank you for the provided solution (manual date conversion). However, I'd like to point out that I am hoping for a configuration or option that allows me to control the time parsing according to a specific time zone. Since I know clearly that the data I read and write, such as "2026-01-31 18:30:40", is associated with a time zone, I expect to be able to parse it directly in that specific time zone, rather than having to handle it manually everywhere.

Thank you for your attention.

**Comment by arp242** [comment]
Go time.Time doesn't have a good way to represent time.Time without a timezone, with the time.Location zero value acting as if it's UTC. This is unfortunate, but not something we can fix here. There does not seem to be a lot of interest in changing this in stdlib at this time (see e.g. https://github.com/golang/go/issues/77398).

The current behaviour of "don't use any timezone" also matches what libpq/psql does, albeit in a way that's a bit confusing, due to time.Time's limitations. It is what it is 🤷

Things like this:

> As others have said before, it makes comparing structures with reflect.DeepEqual impossible because of the timezone.

Is not entirely accurate, because you can do:

	expected := time.Date(2026, 1, 31, 18, 30, 40, 0, time.FixedZone("", 0))

And that works.

What you get from PostgreSQL is "time without a timezone". Which is *not* equal to "time in UTC". It's just that Go's time.Time doesn't have a great way to represent that.

All of this is admittedly rather counter-intuitive. Adding a package level variable might make it easier:

	var WithoutTimezone = time.FixedZone("", 0)

	// even better, but not compatible:
	//var WithoutTimezone = time.FixedZone("without timezone", 0)

But I don't know if it's really worth it?

Adding some sort of option also doesn't really seem worth it. You are requesting a "time without a time zone". I'm not sure if implicitly converting that to something different ("time with a time zone") is really a good idea. All of that will be pq-specific and will cause pq to behave different to libpq or other drivers.

---

In short, there is no great way to handle this and all options come with downsides. I think this is primarily a documentation issue.

- Explicitly specify the timezone in the query:

      select col at time zone 'Asia/Shanghai' from tbl;

  That gets sent as a `timestamptz` and pq handles this correctly.

- Manually create a new time with something like:

      cn, _ := time.LoadLocation("Asia/Shanghai") // Or any other TZ
      t = time.Date(t.Year(), t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), t.Nanosecond(), cn)

      // or:
      t = t.In(cn)
      _, o := have.Zone()
      t = t.Add(-time.Duration(o) * time.Second)


- Use something like https://github.com/golang-sql/civil

- Providing your own time:

	  type TimeCN struct{ time.Time }

	  var cn = func() *time.Location {
		  z, err := time.LoadLocation("Asia/Shanghai")
		  if err != nil {
			  panic(err)
		  }
		  return z
	  }()

	  func (t *TimeCN) Scan(src any) error {
		  tt, ok := src.(time.Time)
		  if !ok {
			  return fmt.Errorf("TimeCN: wrong type: %T", src)
		  }
		  *t = TimeCN{time.Date(tt.Year(), tt.Month(), tt.Day(), tt.Hour(), tt.Minute(), tt.Second(), tt.Nanosecond(), cn)}
		  return nil
	  }
